client/src/nonblocking/tpu_client.rs

```diff
@@ -0,0 +1,470 @@
+            .await?;
+
+        let recent_slots = RecentLeaderSlots::new(start_slot);
+        let slots_in_epoch = rpc_client.get_epoch_info().await?.slots_in_epoch;
+        let leaders = rpc_client
+            .get_slot_leaders(start_slot, LeaderTpuCache::fanout(slots_in_epoch))
+            .await?;
+        let cluster_nodes = rpc_client.get_cluster_nodes().await?;
+        let leader_tpu_cache = Arc::new(RwLock::new(LeaderTpuCache::new(
+            start_slot,
+            slots_in_epoch,
+            leaders,
+            cluster_nodes,
+        )));
+
+        let pubsub_client = if !websocket_url.is_empty() {
+            Some(PubsubClient::new(websocket_url).await?)
+        } else {
+            None
+        };
+
+        let t_leader_tpu_service = Some({
+            let recent_slots = recent_slots.clone();
+            let leader_tpu_cache = leader_tpu_cache.clone();
+            tokio::spawn(async move {
+                Self::run(
+                    rpc_client,
+                    recent_slots,
+                    leader_tpu_cache,
+                    pubsub_client,
+                    exit,
+                )
+                .await
+            })
+        });
+
+        Ok(LeaderTpuService {
+            recent_slots,
+            leader_tpu_cache,
+            t_leader_tpu_service,
+        })
+    }
+
+    pub async fn join(&mut self) {
+        if let Some(t_handle) = self.t_leader_tpu_service.take() {
+            t_handle.await.unwrap().unwrap();
+        }
+    }
+
+    pub fn estimated_current_slot(&self) -> Slot {
+        self.recent_slots.estimated_current_slot()
+    }
+
+    fn leader_tpu_sockets(&self, fanout_slots: u64) -> Vec<SocketAddr> {
+        let current_slot = self.recent_slots.estimated_current_slot();
+        self.leader_tpu_cache
+            .read()
+            .unwrap()
+            .get_leader_sockets(current_slot, fanout_slots)
+    }
+
+    async fn run(
+        rpc_client: Arc<RpcClient>,
+        recent_slots: RecentLeaderSlots,
+        leader_tpu_cache: Arc<RwLock<LeaderTpuCache>>,
+        pubsub_client: Option<PubsubClient>,
+        exit: Arc<AtomicBool>,
+    ) -> Result<()> {
+        let (mut notifications, unsubscribe) = if let Some(pubsub_client) = &pubsub_client {
+            let (notifications, unsubscribe) = pubsub_client.slot_updates_subscribe().await?;
+            (Some(notifications), Some(unsubscribe))
+        } else {
+            (None, None)
+        };
+        let mut last_cluster_refresh = Instant::now();
+        let mut sleep_ms = 1000;
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                if let Some(unsubscribe) = unsubscribe {
+                    (unsubscribe)().await;
+                }
+                drop(notifications);
+                if let Some(pubsub_client) = pubsub_client {
+                    pubsub_client.shutdown().await.unwrap();
+                };
```

Comment:
`notifications` requires a valid reference to `pubsub_client`, so you have to drop `notifications` before moving `pubsub_client`

Reply:
maybe make this a comment to document the sneaky ordering requirement for future code readers